service/storage_service.cc

```diff
@@ -1906,48 +1908,171 @@ future<> storage_service::do_stop_ms() {
+    }
+
+    future<> stop() noexcept {
+        co_await stop_heartbeat_updater();
+    }
+
+    // Caller should set the required req members before prepare
+    future<> prepare(node_ops_cmd cmd) noexcept {
+        return send_to_all(cmd);
+    }
+
+    void start_heartbeat_updater(node_ops_cmd cmd) {
+        if (heartbeat_updater_done_fut) {
+            on_internal_error(slogger, "heartbeat_updater already started");
+        }
+        heartbeat_updater_done_fut = heartbeat_updater(cmd);
+    }
+
+    future<> stop_heartbeat_updater() noexcept {
+        if (heartbeat_updater_done_fut) {
+            as.request_abort();
+            co_await *std::exchange(heartbeat_updater_done_fut, std::nullopt);
+        }
+    }
+
+    future<> done(node_ops_cmd cmd) noexcept {
+        co_await stop_heartbeat_updater();
+        co_await send_to_all(cmd);
     }
-    slogger.info("{}[{}]: Started heartbeat_updater", ops, uuid);
-    while (!(*heartbeat_updater_done)) {
-        auto req = node_ops_cmd_request{cmd, uuid, {}, {}, {}};
+
+    future<> abort(node_ops_cmd cmd) noexcept {
+        co_await stop_heartbeat_updater();
+        co_await send_to_all(cmd);
+    }
+
+    future<> abort_on_error(node_ops_cmd cmd, std::exception_ptr ex) noexcept {
+        slogger.error("{}[{}]: operation failed, sync_nodes={}: {}", desc, uuid(), sync_nodes, ex);
         try {
-          co_await coroutine::parallel_for_each(nodes, [this, ops, uuid, &req] (const gms::inet_address& node) {
-            return _messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([ops, uuid, node] (node_ops_cmd_response resp) {
-                slogger.debug("{}[{}]: Got heartbeat response from node={}", ops, uuid, node);
-                return make_ready_future<>();
-            });
-          });
+            co_await abort(cmd);
         } catch (...) {
-            auto ep = std::current_exception();
-            slogger.warn("{}[{}]: Failed to send heartbeat: {}", ops, uuid, ep);
+            slogger.warn("{}[{}]: abort failed while handling error, sync_nodes={}: {}. Ignoring", desc, uuid(), sync_nodes, std::current_exception());
+        }
+        co_await coroutine::return_exception_ptr(std::move(ex));
+    }
+
+    future<> send_to_all(node_ops_cmd cmd) {
+        req.cmd = cmd;
+        sstring op_desc = format("{}", cmd);
+        bool is_prepare = op_desc.find("prepare") != sstring::npos;
+        slogger.info("{}[{}]: Started {}, sync_nodes={}", desc, uuid(), op_desc, sync_nodes);
+        co_await coroutine::parallel_for_each(sync_nodes, [&] (const gms::inet_address& node) -> future<> {
+            if (nodes_unknown_verb.contains(node) || nodes_down.contains(node) ||
+                    (nodes_failed.contains(node) && op_desc != "abort")) {
+                // Note that we still send to failed nodes that 
```

Comment:
The comments here are not complete.

Reply:
It looks like it is not updated in latest version?

Reply:
Fixed in v4